File: Source/Extensions/Tessa.Extensions.Default.Client/Forums/OpenTopicOnDoubleClickExtension.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Forums;
using Tessa.Platform;
using Tessa.Platform.Runtime;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Controls;
using Tessa.UI.Views;

namespace Tessa.Extensions.Default.Client.Forums
{
    /// <remarks>
    /// У расширения есть конфигуратор <see cref="OpenTopicOnDoubleClickExtensionConfigurator"/>
    /// </remarks>
    public sealed class OpenTopicOnDoubleClickExtension :
        IWorkplaceViewComponentExtension
    {
        private readonly ISession session;
        private readonly IUIHost uiHost;

        public OpenTopicOnDoubleClickExtension(ISession session, IUIHost uiHost)
        {
            this.session = session;
            this.uiHost = uiHost;
        }


        public void Clone(IWorkplaceViewComponent source, IWorkplaceViewComponent cloned, ICloneableContext context)
        {
        }

        public void Initialize(IWorkplaceViewComponent model)
        {
            if (this.session.ApplicationID == ApplicationIdentifiers.TessaAdmin
                || model.RefSection is not null)
            {
                // в TessaAdmin не будем ничего менять, т.к. там предпросмотр представлений
                // в режиме выборки не нужно создавать новую карточку.
                return;
            }
            model.DoubleClickAction = new DoubleClickAction(this.uiHost);
        }

        public void Initialized(IWorkplaceViewComponent model)
        {
        }
        
        private sealed class DoubleClickAction : OpenCardDoubleClickAction
        {
            private readonly IUIHost uiHost;
            public DoubleClickAction(IUIHost uiHost)
            {
                this.uiHost = uiHost;
            }
            protected override async Task OpenCardAsync(Guid cardID, string displayValue, IUIContext context, ViewDoubleClickInfo info)
            {
                var cancellationToken = CancellationToken.None;
                
                var selectedRow = context.ViewContext.SelectedRow;
                if (selectedRow != null)
                {
                    if(selectedRow.TryGetValue("TopicID", out var topicID) && 
                       selectedRow.TryGetValue("TypeID", out var typeID))
                    {
                        using ISplash splash = TessaSplash.Create(TessaSplashMessage.OpeningCard);
                        await this.uiHost.OpenCardAsync(cardID, 
                            options: new OpenCardOptions
                            {
                                DisplayValue = displayValue,
                                UIContext = context,
                                Splash = splash,
                                CardModifierActionAsync = openingContext => CardModifierActionAsync(openingContext, topicID, typeID)
                            },
                            cancellationToken: cancellationToken);

                        /* TODO вывести ошибку пользователю
                        if (objContext == null)
                        {
                           
                        }*/
                    }
                }
            }
            
            
            private static ValueTask CardModifierActionAsync(ICardEditorOpeningContext cardEditorCreationContext, object topicID, object typeID)
            {
                cardEditorCreationContext.Card.Info.Add(ForumHelper.TopicIDKey, topicID);
                cardEditorCreationContext.Card.Info.Add(ForumHelper.TopicTypeIDKey, typeID);
                return new ValueTask();
            }
        }
    }
}

File: WebClient SDK/src/default/forums/openTopicOnDoubleClickExtension.ts
import { ForumHelper } from 'tessa/forums';
import { showLoadingOverlay } from 'tessa/ui';
import { CardEditorOpeningContext } from 'tessa/ui/cards';
import { openCard } from 'tessa/ui/uiHost';
import { DoubleClickInfo, openCardDoubleClickAction } from 'tessa/ui/views';
import { WorkplaceViewComponentExtension } from 'tessa/ui/views/extensions';
import { IWorkplaceViewComponent } from 'tessa/ui/views/workplaceViewComponent';

export class OpenTopicOnDoubleClickExtension extends WorkplaceViewComponentExtension {
  public getExtensionName(): string {
    return 'Tessa.Extensions.Default.Client.Forums.OpenTopicOnDoubleClickExtension';
  }

  public initializeSettings(model: IWorkplaceViewComponent): void {
    model.doubleClickAction = async (info: DoubleClickInfo) => {
      await openCardDoubleClickAction(info, async (cardId, displayValue, context) => {
        const selectedRow = info.context.viewContext?.selectedRow;
        if (selectedRow) {
          const topicId = selectedRow.get('TopicID');
          const typeId = selectedRow.get('TypeID');
          if (topicId && typeId) {
            await showLoadingOverlay(async () => {
              await openCard({
                cardId,
                displayValue,
                context,
                cardModifierAction: openingContext =>
                  OpenTopicOnDoubleClickExtension.cardModifierAction(
                    openingContext,
                    topicId,
                    typeId
                  )
              });
            });
          }
        }
      });
    };
  }

  private static async cardModifierAction(
    openingContext: CardEditorOpeningContext,
    topicId: string,
    typeId: string
  ) {
    openingContext.card.info[ForumHelper.TopicIDKey] = topicId;
    openingContext.card.info[ForumHelper.TopicTypeIDKey] = typeId;
  }
}
